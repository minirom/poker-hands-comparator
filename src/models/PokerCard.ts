import CardRanks from '@/constants/CardRanks';

export class PokerCard {
	public readonly value: string;
	public readonly suit: string;
	public readonly rank: number;

	constructor(card: string) {
		if (card.length > 2 || !RegExp(/[2-9TJQKA]{1}[SHDC]{1}$/).test(card)) {
			throw new Error('Wrong card');
		}
		this.value = card[0];
		this.suit = card[1];
		this.rank = CardRanks.indexOf(this.value);
	}

	public toString(): string {
		return `${this.value === 'T' ? 10 : this.value}${this.suit}`;
	}
}
