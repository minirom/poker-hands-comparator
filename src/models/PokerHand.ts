import { PokerCard } from '@/models/PokerCard';
import { PokerRanks } from '@/enums/PokerRanks';

const SPACE = ' ';

const WEIGHT = [1000, 20, 1, 0.01, 0.0001];

export class PokerHand {
	public readonly cards: PokerCard[];
	public readonly rank: number;
	public readonly suits: string[];
	private readonly occurences: Map<number, number>;

	constructor(cards: string) {
		if (!RegExp(/([2-9TJQKA]{1}[SHDC]{1}\s){4}([2-9TJQKA]{1}[SHDC]{1}){1}$/).test(cards)) {
			throw new Error('Wrong hand');
		}
		this.cards = cards
			.split(SPACE)
			.map((cardValue: string) => new PokerCard(cardValue))
			.sort()
			.sort((cardA: PokerCard, cardB: PokerCard) => cardA.rank - cardB.rank);

		this.suits = this.cards
			.map((card) => card.suit)
			.filter((value, index, array) => array.indexOf(value) === index)
			.sort();

		this.occurences = new Map<number, number>();
		this.countOccurrences();

		this.rank = this.computeRank();
	}

	/**
	 * Check if hand is flush
	 * @returns {boolean} true if flush, false otherwise
	 */
	public isFlush(): boolean {
		return this.suits.length === 1;
	}

	/**
	 * Check if hand is straight
	 * @returns {boolean} true if straight, false otherwise
	 */
	public isStraight(): boolean {
		for (let i = 1; i < this.cards.length; i++) {
			if (this.cards[i].rank - 1 !== this.cards[i - 1].rank) {
				return false;
			}
		}

		return true;
	}

	/**
	 * Check if hand has an ace
	 * @returns {boolean} true if contains ace, false otherwise
	 */
	public hasAce(): boolean {
		return this.cards.map((card: PokerCard) => card.value).includes('A');
	}

	/**
	 * Check if hand is straight flush
	 * @returns {boolean} true if straight flush, false otherwise
	 */
	public isStraightFlush(): boolean {
		return this.isFlush() && this.isStraight();
	}

	/**
	 * Check if hand is royal flush
	 * @returns {boolean} true if royal flush, false otherwise
	 */
	public isRoyalFlush(): boolean {
		return this.isStraightFlush() && this.hasAce();
	}

	/**
	 * Check if hand is four of a kind
	 * @returns {boolean} true if four of a kind, false otherwise
	 */
	public isFourOfAKind(): boolean {
		return this.occurences.size === 2 && Math.max(...Array.from(this.occurences.values()), 0) === 4;
	}

	/**
	 * Check if hand is full house
	 * @returns {boolean} true if full house, false otherwise
	 */
	public isFullHouse(): boolean {
		return this.isThreeOfAKind() && this.occurences.size === 2;
	}

	/**
	 * Check if hand is three of a kind
	 * @returns {boolean} true if three of a kind, false otherwise
	 */
	public isThreeOfAKind(): boolean {
		return Math.max(...Array.from(this.occurences.values()), 0) === 3;
	}

	/**
	 * Check if hand has two pairs
	 * @returns {boolean} true if two pairs, false otherwise
	 */
	public isTwoPairs(): boolean {
		return this.hasPair() && this.occurences.size === 3;
	}

	/**
	 * Check if hand is pair
	 * @returns {boolean} true if flush, false otherwise
	 */
	public isPair(): boolean {
		return this.hasPair() && this.occurences.size === 4;
	}

	public toString(): string {
		return this.cards.join(', ');
	}

	/**
	 * Check if hand has a pair
	 * @returns {boolean} true if contains a pair, false otherwise
	 */
	private hasPair(): boolean {
		return Math.max(...Array.from(this.occurences.values()), 0) === 2;
	}

	/**
	 * Compute a rank of the hand to be able to know if the is hand wins
	 * @returns {number} rank of the hand
	 */
	private computeRank(): number {
		if (this.isRoyalFlush()) {
			return PokerRanks.ROYAL_FLUSH;
		}
		if (this.isStraightFlush()) {
			return PokerRanks.STRAIGHT_FLUSH + this.cards[4].rank;
		}
		if (this.isFourOfAKind()) {
			return PokerRanks.FOUR_OF_A_KIND + this.computeHigherCardsRank();
		}
		if (this.isFullHouse()) {
			return PokerRanks.FULL_HOUSE + this.computeHigherCardsRank();
		}
		if (this.isFlush()) {
			return PokerRanks.FLUSH + this.computeHigherCardsRank();
		}
		if (this.isStraight()) {
			return PokerRanks.STRAIGHT + this.cards[4].rank;
		}
		if (this.isThreeOfAKind()) {
			return PokerRanks.THREE_OF_A_KIND + this.computeHigherCardsRank();
		}
		if (this.isTwoPairs()) {
			return PokerRanks.TWO_PAIRS + this.computeHigherCardsRank();
		}
		if (this.isPair()) {
			return PokerRanks.PAIR + this.computeHigherCardsRank();
		}

		return PokerRanks.HIGH_CARD + this.computeHigherCardsRank();
	}

	/**
	 * Count occurrences of each cards
	 */
	private countOccurrences(): void {
		this.cards.forEach((card: PokerCard) => {
			const occurences = this.occurences.get(card.rank);
			if (occurences) {
				this.occurences.set(card.rank, occurences + 1);
			} else {
				this.occurences.set(card.rank, 1);
			}
		});
	}

	/**
	 * Compute rank by giving a weight for best cards
	 * @returns {number} rank from higher cards
	 */
	private computeHigherCardsRank(): number {
		const values = Array.from(this.occurences).sort((a, b) => b[1] - a[1]);
		return values
			.map((value, index) => value[0] * value[1] * WEIGHT[index])
			.reduce((acc, value) => acc + value, 0);
	}
}
