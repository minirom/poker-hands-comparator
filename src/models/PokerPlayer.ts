import { PokerHand } from '@/models/PokerHand';

export class PokerPlayer {
	public hand: PokerHand;

	constructor(public readonly name: string, cards: string) {
		this.hand = new PokerHand(cards);
	}

	public switchCards(cards: string) {
		this.hand = new PokerHand(cards);
	}
}
