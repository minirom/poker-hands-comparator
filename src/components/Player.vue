<template>
	<div>
		{{ player.name }} : {{ player.hand }}
		<button @click="$emit('delete', player)">Delete</button>
	</div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';
import { PokerPlayer } from '@/models/PokerPlayer';

@Component
export default class Player extends Vue {
	@Prop() private player!: PokerPlayer;
}
</script>
