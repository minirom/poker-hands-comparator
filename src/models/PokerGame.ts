import { PokerPlayer } from '@/models/PokerPlayer';
import allCards from '@/constants/AllCards';
import { PokerService } from '@/services/PokerService';

export class PokerGame {
	public pokerPlayers: PokerPlayer[];
	public winners?: PokerPlayer[];
	private cards?: string[];

	constructor(names: string[]) {
		if (allCards.length / names.length < 5) {
			throw new Error('Too many players');
		}
		this.randomSortCards();
		this.pokerPlayers = names.map((name) => new PokerPlayer(name, this.generateRandomHand()));
		this.findWinners();
	}

	/**
	 * Generate new game
	 */
	public generateNewGame(): void {
		this.randomSortCards();
		this.pokerPlayers.forEach((player: PokerPlayer) =>
			player.switchCards(this.generateRandomHand())
		);
		this.findWinners();
	}

	/**
	 * Add player
	 * @param {string} name: name of new player
	 */
	public addPlayer(name: string): void {
		if (allCards.length / (this.pokerPlayers.length + 1) < 5) {
			throw new Error('It will have too many players in the game');
		}
		if (name.trim().length === 0) {
			throw new Error('Name is empty');
		}

		const filteredName = this.pokerPlayers.filter(
			(pokerPlayer: PokerPlayer) => pokerPlayer.name === name
		);
		if (filteredName.length > 0) {
			throw new Error('Name already exists');
		}
		this.pokerPlayers.push(new PokerPlayer(name, this.generateRandomHand()));
		this.findWinners();
	}

	/**
	 * Remove player
	 * @param {PokerPlayer} player : player to remove
	 */
	public removePlayer(player: PokerPlayer): void {
		const indexPlayer = this.pokerPlayers.findIndex(
			(pokerPlayer: PokerPlayer) => pokerPlayer === player
		);
		if (indexPlayer >= 0) {
			player.hand.cards.forEach((card) => {
				// @ts-ignore
				this.cards.push(`${card.value}${card.rank}`);
			});
			this.randomSortCards();
			this.pokerPlayers.splice(indexPlayer, 1);
		}
		this.findWinners();
	}

	/**
	 * Find winners
	 */
	private findWinners() {
		this.winners = PokerService.findWinners(this.pokerPlayers);
	}

	/**
	 * Sort cards randomly
	 */
	private randomSortCards(): void {
		this.cards = [...allCards];
		this.cards.sort(() => Math.random() - 0.5);
	}

	/**
	 * Generate random hand
	 */
	private generateRandomHand(): string {
		const cards = [];
		for (let i = 0; i < 5; i++) {
			// @ts-ignore
			const index = Math.floor(Math.random() * (this.cards.length - 1));
			// @ts-ignore
			cards.push(this.cards[index]);
			// @ts-ignore
			this.cards.splice(index, 1);
		}
		return cards.join(' ');
	}
}
