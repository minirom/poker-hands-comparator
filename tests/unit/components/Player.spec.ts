import { mount } from '@vue/test-utils';
import createFactory from '../../utils/createFactory';
import Player from '@/components/Player.vue';
import { PokerPlayer } from '@/models/PokerPlayer';

// @ts-ignore
const factory = createFactory(({ props }) =>
	mount(Player, {
		propsData: {
			...props
		}
	})
);

test('Player component', () => {
	describe('should create poker player', () => {
		const wrapper = factory({ props: { player: new PokerPlayer('1', 'AS KS QS JS TS') } });
		expect(wrapper).toMatchSnapshot();
	});
});
