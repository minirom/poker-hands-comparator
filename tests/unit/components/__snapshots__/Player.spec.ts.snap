// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player component 1`] = `
<div>
  1 : 10S, JS, QS, KS, AS
  <button>Delete</button></div>
`;
